Add `cirq.eject_phased_paulis` transformer to replace `cirq.EjectPhasedPaulis`

tanujkhattar opened the pull request:
- Part of https://github.com/quantumlib/Cirq/issues/4722
- Follows the new Transformer API https://github.com/quantumlib/Cirq/issues/4483
- Supports no compile tags NoCompile Tag for optimizers https://github.com/quantumlib/Cirq/issues/4253


Breaking change: `cirq.EjectPhasedPaulis` optimizer will now remove empty moments by default.

MichaelBroughton commented on cirq-core/cirq/transformers/eject_phased_paulis_test.py:
```diff
@@ -0,0 +1,560 @@
+    )
+
+
+def test_toggles_measurements():
```
Can we add additional tests here for CCO and MeasurementKey interactions ?

tanujkhattar replied:
done

MichaelBroughton reviewed (commented):
A few nits and then two bigger feedback points:

1. We should probably have this work with PhasedXZ
2. Is there a change in behavior on the old transformer that we should fix ?

tanujkhattar commented on cirq-core/cirq/optimizers/eject_phased_paulis_test.py:
```diff
@@ -26,9 +26,11 @@ def assert_optimizes(
+        opt = cirq.EjectPhasedPaulis(eject_parameterized=eject_parameterized)
 
     circuit = before.copy()
+    expected = cirq.drop_empty_moments(expected)
```
Yes, that's correct. Though I'm not aware of a use case where people would like to keep empty moments in their circuit, so it's a relatively safe breaking change. 

The common workflow is the call `drop_empty_moments` after `EjectPhasedPaulis` (as seen in existing usages of this transformer throughout cirq -- for example: `optimized_for_sycamore` method, `spin_echoes.ipynb` tutorial etc.)

MichaelBroughton replied:
Ok, I think we should be good to make that little change then, but I think we should still mark this PR as a breaking change and document it for posterity's sake.

tanujkhattar replied:
done

tanujkhattar commented on cirq-core/cirq/transformers/analytical_decompositions/two_qubit_to_cz.py:
```diff
@@ -23,9 +23,9 @@
 
 from cirq import ops, linalg, protocols, circuits
 from cirq.transformers.analytical_decompositions import single_qubit_decompositions
+from cirq.transformers.eject_phased_paulis import eject_phased_paulis
```
Module import causes problems in this case because `eject_phased_paulis` is the module name as well as the function name. I've followed this pattern for other imports from the transformer package as well, like `eject_z`. 

I can change it to a lazy import if you have strong preference.

MichaelBroughton replied:
Ahhh, that's fine, no strong preference.

tanujkhattar commented on cirq-core/cirq/transformers/eject_phased_paulis.py:
```diff
@@ -0,0 +1,330 @@
+    atol: float = 1e-8,
+    eject_parameterized: bool = False,
+) -> 'cirq.Circuit':
+    """Transformer pass to push X, Y, and PhasedX gates towards the end of the circuit.
```
Added support for `PhasedXZ` gates with z_exponent = 0 (which are equivalent to PhasedX gates). Added a test which demonstrates how to deal with a general `PhasedXZGate`, i.e. 
run `cirq.eject_z(cirq.eject_phased_paulis(cirq.eject_z(c)))`. 
1. `cirq.eject_z(c)` : This ejects Z gates and ensures that the `z_component` of `PhasedXZGate` is 0. 
2. `cirq.eject_phased_paulis(c)`: This ejects `PhasedXPowGates`, potentially leaving z phased from CZ phase kickback, which will then be removed by the 3rd `cirq.eject_z()`.

MichaelBroughton replied:
Awesome, could we also support the case where axis_exponent=0 and x_exponent=0 but z_exponent can be whatever we want ? This seems like it might be able to be worked into `_try_get_known_z_half_turns` without much difficulty ?

tanujkhattar commented on cirq-core/cirq/transformers/eject_phased_paulis.py:
```diff
@@ -0,0 +1,330 @@
+    As the gates get pushed, they may absorb Z gates, cancel against other
+    X, Y, or PhasedX gates with exponent=1, get merged into measurements (as
+    output bit flips), and cause phase kickback operations across CZs (which can
+    then be removed by the EjectZ optimization).
```
done

tanujkhattar commented on cirq-core/cirq/optimizers/eject_phased_paulis.py:
```diff
@@ -15,27 +15,11 @@
-        self.inline_intos: List[Tuple[int, ops.Operation]] = []
-        self.insertions: List[Tuple[int, ops.Operation]] = []
+from cirq import circuits, transformers
+from cirq._compat import deprecated_class
```
done

tanujkhattar commented:
@MichaelBroughton Addressed Feedback, ready for another look.

MichaelBroughton commented on cirq-core/cirq/transformers/eject_phased_paulis.py:
```diff
@@ -0,0 +1,342 @@
+    atol: float = 1e-8,
+    eject_parameterized: bool = False,
+) -> 'cirq.Circuit':
+    """Transformer pass to push X, Y, and PhasedX gates towards the end of the circuit.
```
Might want to also mention PhasedXZGates in the topline of the docstring.

MichaelBroughton commented on cirq-core/cirq/transformers/eject_phased_paulis_test.py:
```diff
@@ -0,0 +1,590 @@
+    # CCOs
+    assert_optimizes(
+        before=quick_circuit(
+            [cirq.PhasedXPowGate(phase_exponent=0.25).on(a)],
+            [cirq.measure(a, key="m")],
+            [cirq.X(b).with_classical_controls("m")],
```
What would be the expected output for something like ?
```python3
[cirq.PhasedXPowGate(phase_exponent=0.25).on(a)],
[cirq.measure(a, key="m")],
[cirq.X(b).with_classical_controls("m")],
[cirq.measure(b, key='m2')]
```
?

tanujkhattar replied:
`cirq.X(b).with_classical_controls` will be unaffected because it's not a unitary operation. So, the output will be:
```
a: ───!M─────────────────
      ║
b: ───╫────X───M('m2')───
      ║    ║
m: ═══@════^═════════════
```

MichaelBroughton commented on cirq-core/cirq/transformers/eject_phased_paulis.py:
```diff
@@ -0,0 +1,342 @@
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    atol: float = 1e-8,
```
This seems a little tight, floats are usually only good up to 1e-6 or so.

tanujkhattar replied:
1e-8 tolerance is standard across the codebase, try running `grep -RI '1e-8' --include \*.py`.

MichaelBroughton reviewed (approved):
LGTM, two concerns now are:

1. Can we accomodate the PhasedXZ case where axis_exponent=0 and x_xeponent=0 (aka just a Z).
2. Can we double check that CCO behave themselves when they are the operations that are getting moved (it's good that things work when they aren't getting moved) ?

tanujkhattar commented:
1. Added support for PhasedXZ gates with only z_exponent as the non-zero angle. Also updated docstrings. 
2. CCO's will not get moved because `op.gate` is None when `op` is a `CCO`, so the operation would be an "unknown" operation and will act as a blocker. Same is true for other existing transformers as well. Adding support for running transformers on the gates inside CCOs is currently being discussed in https://github.com/quantumlib/Cirq/issues/4923 and I'd prefer to not block this PR for the same.
